src/main/kotlin/com/github/julianhille/sqldelight/jscipherdriver/JsSqlCipherDriver.kt

```diff
@@ -0,0 +1,207 @@
+  private var transaction: Transacter.Transaction? = null
+
+
+  init {
+    if (configuration.journalMode ) db.pragma("journal_mode = WAL")
+    if (configuration.key?.isNotBlank() == true) db.pragma("journal_mode = WAL")
+    migrateIfNeeded(configuration.create, configuration.upgrade, configuration.schema.version)
+  }
+
+  override fun close() {
+    db.close()
+  }
+
+  override fun execute(identifier: Int?, sql: String, parameters: Int, binders: (SqlPreparedStatement.() -> Unit)?) {
+    createOrGetStatement(identifier, sql).run {
+      bind(binders)
+      run()
+    }
+  }
+
+  private fun Statement.bind(binders: (SqlPreparedStatement.() -> Unit)?) = binders?.let {
+    val bound = SqlJsCipherStatement()
+    binders(bound)
+    println(bound.parameters)
+    bind(bound.parameters.toTypedArray())
+  }
+
+  override fun executeQuery(
+    identifier: Int?,
+    sql: String,
+    parameters: Int,
+    binders: (SqlPreparedStatement.() -> Unit)?
+  ): SqlCursor {
+    return createOrGetStatement(identifier, sql).run {
+      bind(binders)
+      SqlJsCipherCursor(iterate())
+    }
+  }
+
+  override fun newTransaction(): Transacter.Transaction {
+    val enclosing = transaction
+    val transaction = Transaction(enclosing)
+    this.transaction = transaction
+    if (enclosing == null) {
+      db.exec("BEGIN TRANSACTION")
+    }
+    return transaction
+  }
+
+  override fun currentTransaction() = transaction
+
+  private inner class Transaction(
+    override val enclosingTransaction: Transacter.Transaction?
+  ) : Transacter.Transaction() {
+    override fun endTransaction(successful: Boolean) {
+      if (enclosingTransaction == null) {
+        if (successful) {
+          db.exec("END TRANSACTION")
+        } else {
+          db.exec("ROLLBACK TRANSACTION")
+        }
+      }
+      transaction = enclosingTransaction
+    }
+  }
+
+  private fun createOrGetStatement(identifier: Int?, sql: String): Statement = db.prepare(sql)
+
+  fun getVersion(): Int = (db.pragma("user_version") as? Double)?.toInt() ?: 0
+
+  private fun setVersion(version: Int): Unit {
+    db.pragma("user_version=${version}")
+  }
+
+  fun migrateIfNeeded(
+    create: (SqlDriver) -> Unit,
+    upgrade: (SqlDriver, Int, Int) -> Unit,
+    version: Int
+  ) {
+    val initialVersion = getVersion()
+    println("soimething mumble mumble $initialVersion")
+    println("soimething mumble mumble $db")
+    println("soimething mumble mumble $transaction")
+    val driver = this
+    if (initialVersion == 0) {
+      transaction.run {
+        create(driver)
+        setVersion(version)
+      }
+    } else if (initialVersion != version) {
+      if (initialVersion > version) {
+          throw IllegalStateException("Database version $initialVersion newer than config version $version")
+        }
+        transaction.run {
+          upgrade(driver, initialVersion, version)
+          setVersion(version)
+        }
+    }
+  }
```

Comment:
Indentation

Reply:
rereview this one please as i changed the block completely